<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Resources for lab developers</title>

<script src="site_libs/header-attrs-2.10/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<link rel="shortcut icon" href="https://github.com/FredHutch/data/raw/main/resources/images/favicon.ico" />
 <!--- go to https://favicon.io/favicon-converter/ to upload an image to make a new favicon.io. You will need to replace the current favicon.io image with the one in the downloaded directory from the website. The current image is in the resources/images/ directory --->

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>





<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img src='resources/images/FredHutch_h_tag_4col_cmyk.jpg' width = '200px'/></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    About Code Review
  </a>
</li>
<li>
  <a href="lab_leaders.html">
    <span class="fa fa-person-chalkboard"></span>
     
    For Lab Leaders
  </a>
</li>
<li>
  <a href="lab_managers.html">
    <span class="fa fa-users-between-lines"></span>
     
    For Lab Managers
  </a>
</li>
<li>
  <a href="lab_developers.html">
    <span class="fa fa-users-gear"></span>
     
    For Lab Developers
  </a>
</li>
<li>
  <a href="more_resources.html">
    <span class="fa fa-book"></span>
     
    So many resources!
  </a>
</li>
<li>
  <a href="https://data.fredhutch.org">
    <span class="fa fa-code-merge"></span>
     
    FH DaSL
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Resources for lab developers</h1>

</div>


<p>As the direct author of the code you will have the most intimate knowledge of the code and project details. This means that your ability to communicate with others in your lab is critical to the smoothness and success of the project.</p>
<div id="general-tips" class="section level2">
<h2>General tips:</h2>
<ul>
<li>Learning to data science can be a steep learning curve. <strong>Be patient with yourself!</strong> Anyone who has ever written code has definitely wanted to pull their hair out at one time or another when they’ve gotten stuck on a tricky issue.</li>
<li>In respect to this, we recommend trying to <strong>get plugged into coding groups</strong>. Not only can they be helpful for helping you get unstuck, but also just helpful for support in general or for learning tips for how to improve your code.</li>
<li><strong>Seek out resources</strong> that can help give you the basics of coding. Take a look at our <a href="./more_resources.html">pretty long list</a>. Think about how you best learn. Do you like <a href="./more_resources.html#Free-online-resources">reading through material on your own</a>, or do you prefer a more formal class setting like what a <a href="./more_resources.html#training-networks">workshop</a> might provide? If you want a mix of both structure and work-at-your-own pace, <a href="./more_resources.html#learn-on-your-own-courses">online courses</a> may be the way to start.</li>
<li><strong>Data science is best performed as a team effort</strong>. There’s no such thing as the “lone genius” in data science. Don’t be afraid to ask others for help! Sometimes you just explaining the problem will help you to see a solution!</li>
<li>It is a skill to know when to ask for help and when you should continue to plug away at a problem on your own. With the help and guidance of your lab manager and/or lab leader, you can work on striking the right balance. On one hand, if you are stuck on a problem that you are making no progress on it is an inefficient use of everyone’s time and energy for you to continue to be frustrated – <strong>definitely reach out for help</strong>! Others want you and your project to succeed and are generally happy to help you!</li>
<li>If you are in a small lab or a lab that is not very programming proficient, you may need to get creative in finding help. At the data science lab, one of our goals will be to help you get connected with others to collaborate with. You can also take a look at <a href="./more_resources.html#groups-for-discussing-code">online and in person coding groups</a>.</li>
</ul>
</div>
<div id="tips-on-debugging" class="section level2">
<h2>Tips on debugging</h2>
<p>So much of data science is formatting and troubleshooting. See this guide for some handy tips for that. Bugs don’t always come in the form of actual error messages either. The worst bugs are often ones that are silent and have messed up your results without you knowing.</p>
<p>We are borrowing from the <a href="https://github.com/AlexsLemonade/training-modules/blob/master/intro-to-R-tidyverse/00b-debugging_resources.md">debugging guide from the Childhood Cancer Data Lab here</a>:</p>
<div id="carefully-read-any-and-all-error-messages" class="section level3">
<h3>1) Carefully read any and all error messages</h3>
<p>This may seem like a silly thing to include as a tip, but it’s very easy to gloss over an error message without actually reading it. Often, R may be telling you exactly what is wrong, but if you don’t take the time to understand what the error message means, you will have trouble fixing the error. Error messages often refer to R terms (e.g.. “argument”, “directory”) so if you need a refresher on what some terms mean, we recommend running through the intro-to-R notebook or going through Hadley Wickham’s R for Data Science</p>
<p>Secondly, realize that just because you don’t receive an error message, doesn’t mean that your code did what you intended it to. You also will need to carefully review your code (and your results) to try to find “silent” bugs (situations where R did exactly what you asked, but you didn’t get what you intended).</p>
</div>
<div id="identify-which-line-and-phrase-of-code-is-the-source-of-the-error." class="section level3">
<h3>2) Identify which line and phrase of code is the source of the error.</h3>
<p>If you ran many lines of code, you may not know which part of your code is the origin of the error message. Isolating the source of the error and trying to better understand your problem should be your first course of action. The best way to determine this, is by running each line, and each phrase by itself, one at a time.</p>
<p>Chunk-out your code and test the individual bits of code. Do you have a lot of lines of code, a lot of arguments, or multiple functions at once? Try each piece by itself to narrow down what piece appears to be the origin of the problem.</p>
</div>
<div id="be-sure-that-the-code-you-think-you-have-run-has-all-successfully-run-and-in-order." class="section level3">
<h3>3) Be sure that the code you think you have run has all successfully run and in order.</h3>
<p>It could be that the problem with your code isn’t that it doesn’t work as it is written, but that you didn’t run it or didn’t run it in the correct order. This should be one of the first things you check, while checking that the objects that you believe should be in your environment, are in your environment.</p>
<p>It’s also good practice to be periodically quitting your current R session and starting a new one, in addition to clearing your R notebook output. If you are encountering problems and haven’t refreshed your R session, you may want to do that before further troubleshooting.</p>
<p>In the course of troubleshooting, you will want to re-run all of your code, perhaps many many many times in order to get to the bottom of the problem.</p>
</div>
<div id="google-your-error-message" class="section level3">
<h3>4) Google your error message</h3>
<p>The main advantage to Googling your errors is that you are likely not the first person to encounter the problem. Certain phrases and terms in the error message will yield more fruitful search results then others.</p>
<p>When you do Google, a few common sources that will probably come up that we recommend looking at are:</p>
<div id="a-stackoverflow" class="section level4">
<h4>a) StackOverflow</h4>
<p>StackOverflow this is a forum where people post questions and problems they encounter in their code.</p>
</div>
<div id="b-github-issues" class="section level4">
<h4>b) GitHub Issues</h4>
<p>People also will post their problems to GitHub issues. Often these are more geared toward fixing problems with the package or software itself, but this is a way to potentially get direct help on an issue from the authors of the package you are using.</p>
</div>
<div id="c-r-bloggers" class="section level4">
<h4>c) R-bloggers</h4>
<p>R-bloggers has examples of R code that you can use to figure out how to construct various analyses. This is a good resource for example code, although it’s format isn’t built for asking exact questions like StackOverflow.</p>
</div>
</div>
<div id="look-at-the-documentation-for-a-function-to-make-sure-you-are-using-it-correctly" class="section level3">
<h3>5) Look at the documentation for a function to make sure you are using it correctly</h3>
<p>Once you’ve better determined the origin of the problem, you should use whatever documentation is available to you regarding the problematic code. When using a new function from a package you are unfamiliar with, it’s worthwhile to skim through the documentation so you know how to properly use the functions. For base R functions, Tidyverse functions, and some Bioconductor packages, the documentation will give you a lot of the information you need. However, you will also likely find that not all documentation is thorough or clear.</p>
<p>As we discussed in intro_to_R notebook, objects have structures and types. Having input that doesn’t match the requirements that a function has can be a common source of errors. Pay special attention to what the documentation says about what kind of input and output the function is designed to use.</p>
<div id="use-the-rstudio-help-bar" class="section level4">
<h4>Use the RStudio help bar</h4>
<p>Here’s a screenshot from the help window in RStudio. Note that here we searched for the levels function. R documentation includes information about what the expected arguments are as well as examples of how to use a function. Note here that this documentation tells us that the input for x is probably a factor. search_bar</p>
<p>For Bioconductor package functions, look at their page on bioconductor.org Like other a lot of other packages we may be using for genomics analysis, DESeq2 is a Bioconductor package. We use DESeq2 in the bulk RNA-seq module.</p>
<p>The documentation on Bioconductor pages have information that can be valuable for troubleshooting. Vignettes can have good example workflows to get started with (can use the browseVignettes function for RStudio to open them). In addition, every Bioconductor package has a PDF reference where all the functions and objects for that package are described. They can take some getting used to, but generally can have helpful information.</p>
</div>
</div>
<div id="google-it-again" class="section level3">
<h3>6) Google it again</h3>
<p>Because it’s unlikely your first attempt at Googling will lead you straight to an answer; this is something you should continue to try with different wordings. Through trial and error, and also Google algorithms learning about what you look for, your search results can eventually lead you to helpful examples and forums.</p>
</div>
<div id="look-at-the-source-code-for-that-function" class="section level3">
<h3>7) Look at the source code for that function</h3>
<p>This should rarely be your first approach to solving a problem, since this approach is difficult and doesn’t always pay off. This approach will require a a bit more practice at reading code, so it may not be the most fruitful approach depending on the readability and complexity of the code.</p>
</div>
<div id="post-to-an-appropriate-forum-on-stackoverflow-or-a-github-issue" class="section level3">
<h3>8) Post to an appropriate forum on StackOverflow or a GitHub Issue</h3>
<p>After you’ve tediously mined the internet for solutions to your problem and still not resolved your problem, you can post your problem to the internet for help.</p>
</div>
</div>
<div id="tips-on-asking-for-help" class="section level2">
<h2>Tips on asking for help</h2>
<p>Asking for help is a skill in itself! You will be able to more successfully receive help from others if you can better communicate the problem from the get go. Read this article for more details. We will cover the basics here.</p>
<p><strong>The best coding help requests include</strong> (paraphrased from the article by Jere Xu):<br />
- <strong>A description of the problem</strong>— Explain what you are doing in the first place and what language and frameworks you are using. Give context to what you are trying to do! - <strong>What code did you use to get this problem</strong> — Provide your exact code and data you used and detail exactly what you did in order to get to the problem - <strong>Expected result</strong> — Describe what the intended result of your code is and maybe even show a real-world example (assuming that what you’re building isn’t completely new). - <strong>Actual result</strong> - Copy and paste the exact error message you are getting or otherwise provide a screenshot of the problem you are seeing. - <strong>Environment</strong> — What operating system and version are you running on? What package managers and libraries are you using?</p>
</div>
<div id="reading-about-code-review" class="section level2">
<h2>Reading about code review</h2>
<ul>
<li>For tips for writing good code and creating reproducible projects: <a href="https://jhudatascience.org/Reproducibility_in_Cancer_Informatics/">Reproducibility in Cancer Informatics</a></li>
<li>For tips for writing pull requests: <a href="https://jhudatascience.org/Adv_Reproducibility_in_Cancer_Informatics/engaging-in-code-review---as-an-author.html">Engaging in Code Review as an Author</a> by Candace Savonen.</li>
<li>For tips for reviewing pull requests: <a href="https://jhudatascience.org/Adv_Reproducibility_in_Cancer_Informatics/engaging-in-code-review---as-a-reviewer.html">Engaging in Code Review as an Reviewer</a> by Candace Savonen.</li>
<li>For more on effective code review: <a href="https://www.freecodecamp.org/news/a-zen-manifesto-for-effective-code-reviews-e30b5c95204a/">A zen manifesto for effective code reviews</a> by Jean-Charles Fabre.</li>
<li>About <a href="https://github.com/AlexsLemonade/training-modules/blob/master/intro-to-R-tidyverse/00c-good-scientific-coding-practices.md">Good scientific coding practices</a> by the Childhood Cancer Data Lab.</li>
<li><a href="https://github.com/AlexsLemonade/training-modules/blob/master/intro-to-R-tidyverse/00b-debugging_resources.md">Tips for debugging code from the Childhood Cancer Data Lab</a></li>
<li>For good practices on code review: <a href="https://smartbear.com/en/learn/code-review/best-practices-for-peer-code-review/">Best practices for Code Review</a> from SmartBear.</li>
</ul>
<p><strong>See our list of <a href="./more_resources.html">recommended resources</a> to get going!</strong></p>
</div>

<hr>
<center>
  <div class="footer">
    <img src="resources/images/FredHutch_v_tag_4col_cmyk.jpg" width=85>
    <br>
    <a href="https://github.com/jhudsl/OTTR_Template/issues/new/choose">File a GitHub issue here</a>.
  </div>
</center>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
